include/revng/RestructureCFG/MarkerCallUtils.h

```diff
@@ -0,0 +1,222 @@
+    if (not MarkerCallF) {
+      // Create or get the marker function declaration, that will be called by
+      // the inserted markers
+      llvm::Type *BlockAddressTy = llvm::Type::getInt8PtrTy(C);
+      auto *FT = llvm::FunctionType::get(llvm::Type::getVoidTy(C),
+                                         { BlockAddressTy },
+                                         false);
+
+      using llvm::Function;
+      using llvm::GlobalValue;
+      MarkerCallF = llvm::cast<
+        llvm::Function>(M->getOrInsertFunction(MarkerFunctionTag.name(), FT)
+                          .getCallee());
+      MarkerCallF->setLinkage(llvm::GlobalValue::ExternalLinkage);
+      MarkerCallF->addFnAttr(llvm::Attribute::OptimizeNone);
+      MarkerCallF->addFnAttr(llvm::Attribute::NoInline);
+      MarkerCallF->addFnAttr(llvm::Attribute::NoMerge);
+      MarkerCallF->addFnAttr(llvm::Attribute::NoUnwind);
+      MarkerCallF->addFnAttr(llvm::Attribute::WillReturn);
+      using llvm::MemoryEffects;
+      MarkerCallF->setMemoryEffects(MemoryEffects::inaccessibleMemOnly());
+
+      // Add the tag to the
+      MarkerFunctionTag.addTo(MarkerCallF);
+    }
+  }
+
+public:
+  // Set the insertion point of the `MarkerCallBuilder`
+  void setInsertPoint(llvm::BasicBlock *NewBB) { BB = NewBB; }
+
+  // If the argument is `nullptr`, it means that we are inserting a marker for
+  void insertMarkerCall(llvm::BasicBlock *BasicBlockTarget = nullptr) {
+
+    // We must have an insertion point
+    revng_assert(BB);
+
+    // We assume that when inserting a `goto` edge, the original block had a
+    // single regular successor on the CFG
+    if (MarkerCallF->getName() == GotoBlockMarkerCallFunctionName) {
+      llvm::Instruction *Terminator = BB->getTerminator();
+      revng_assert(Terminator->getNumSuccessors() == 1);
+    }
+
+    // We always insert the marker as the penultimate instruction in a
+    // `BasicBlock`, regardless of the type of the marker we are inserting
+    llvm::Instruction *Terminator = BB->getTerminator();
+    llvm::IRBuilder<> Builder(Terminator);
+
+    // When creating a `goto_block` marker, we don't have a target block as
+    // argument, and we place a `nullptr` as argument of the marker call
+    if (BasicBlockTarget) {
+      auto *BasicBlockAddressTarget = llvm::BlockAddress::get(BasicBlockTarget);
+      revng_assert(BasicBlockAddressTarget);
+      Builder.CreateCall(MarkerCallF, BasicBlockAddressTarget);
+    } else {
+      llvm::LLVMContext &C = getContext(BB);
+      llvm::PointerType *BlockAddressTy = llvm::Type::getInt8PtrTy(C);
+      auto
+        *NullBasicBlockAddress = llvm::ConstantPointerNull::get(BlockAddressTy);
+      Builder.CreateCall(MarkerCallF, NullBasicBlockAddress);
+    }
+  }
+};
+
+/// Helper method to retrieve the `BasicBlock` target of the marker
+inline llvm::BasicBlock *getMarkerCallTarget(std::string MarkerCallName,
+                                             llvm::BasicBlock *BB) {
```

Comment:
I don't think we need this. We just want to have `getGotoMarkerCall and `getScopeCloserCall`

Reply:
This method disappeared due to all the changes to the code relevant for this PR